pkg/router/template/router.go

```diff
@@ -865,17 +870,30 @@ func generateDestCertKey(config *ServiceAliasConfig) string {
 
 type endpointsCounter func(key string) int32
 
-// getServiceUnits returns a map of service keys to their weights.
-// The requests are loadbalanced among the services referenced by the route.
+// getServiceUnits returns a map of service keys to their weights
+// and a map of service keys to number of endpoints in the service.
+//
+// The requests are load balanced among the services referenced by the route.
 // The weight (0-256, default 1) sets the relative proportions each
-// service gets (weight/sum_of_weights fraction of the requests).
+// service gets (weight/sum_of_weights) fraction of the requests.
 // When the weight is 0 no traffic goes to the service. If they are
 // all 0 the request is returned with 503 response.
-// For each service, the requests are distributed among the endpoints.
-// Each endpoint gets weight/numberOfEndpoints portion of the requests.
+// For each service, the requests are distributed among the service's
+// endpoints.
+//
+// The service weight is distributed among the endpoints. Since the
+// minimum weight is 1 it is possible that there are too many endpoints
+// for the service weight. E.g, service weight 5 and 9 endpoints.
+// In this case 5 endpoints will get 1 and 4 will get 0. Otherwise,
```

Comment:
This seems incorrect.  We should never get into this state.  Behaving this way at this level seems wrong.

Reply:
We need to allow assigning weight 0 to pods. This is useful for stateful pods to handle existing sessions while they are draining (trello card: https://trello.com/c/qHZm47LV).

Reply:
Weight > 0 means "send some traffic to this service".  A user who is setting all weight 1 is not telling us to ignore certain backends, they're requesting even balancing.  Weight is about relative balance, and we should increase the weight of the high end until the ratios are correct.  

If we have weights 2 and 1, and number endpoints 1 and 10, then the first pod should have weight 20 and all the rest should have 1.

Reply:
Basically, the weight from the end user is not the value we are carving up.  We are distributing shares based on proportional weight, and clamping at 1 and whatever the high end is.

Reply:
@smarterclayton correct. Only that the high end is bounded. haproxy limits the weights to integer between 1 and 256. That upper bound forces us to be sparse at the lower bound.

Reply:
I would increase the high range to 256 before clamping.  Giving pods weight zero is wrong - the autoscaler has the responsibility to make that decision, not the router.  It's acceptable to be weight imbalanced at the extremes (worst case is equal weighting between one pod with weight 256 and +256 pods with weight 1 - that's the users problem, not ours).

Reply:
Concrete example.  User has two services, one with autoscaler and one without.  They set weight 50/50 and walk away.  Service one has 5 pods, service two has 5 pods to start.  Traffic grows, autoscaler grows second service to accommodate traffic.  We should err on the side of giving traffic to all service two pods, no matter how unbalanced it gets.

Reply:
Fair point. That is what we do today - not be partial to some pods of the same service while accommodating as much of AB ratio as possible. 
This PR was supposed to provide the stickler option. Do you think we should not give that option at all?
Either which way, I feel we should utilise Phil's excellent examples for the docs.

Reply:
@smarterclayton "Weight > 0 means "send some traffic to this service".

The service is not actually what is being balanced. Balancing is done by weight at the endpoints, not the service. When a user wants to distribute requests A/B the collection of endpoints needs to be weighted to get the desired service ratio. This is hard for the user to figure out, instead we go for an easier for the user to understand approach, balancing weights among services. 

Suppose the user wants 1% to A and 99% to B and there are 300 A endpoints and 300 B endpoints. All A endpoints could get 2 and all B endpoints could get 200 which would distribute load to all endpoints and achieve the desired service balance. The current algorithm doesn't do that. I will fix it.

Since weights are limited to the range 1-256 with 0 meaning no traffic. There will always be corner cases, which we will seldom encounter, where some endpoints will have to be set to 0 to establish the desired balance. Imagine an unreasonable, but possible, configuration of 1% A with 600 endpoints and 99% B with 2 endpoints. To achieve the desire balance the B endpoints would get 256, 512 A endpoints would get 1 and 88 would get 0. Arg!